[Tabs] MDCTabBarViewController - like UITabBarViewController

DavidPhillipOster opened the pull request:
but uses Material Design's MDCTabBar

Usable. In future pull requests I'll add documentation, an Interface Builder example, and a Swift unit test.

cr/154488014

Commit:
MDCTabBarViewController - like UITabBarViewController, but uses Material Design's MDCTabBar

Commit:
MDCTabBarViewController Interface Builder example

It's integrated into the Catalog app as a Tab example. Also demonstrates moving the tab bar from the unselected state to the selected state. Note: this reveals a bug with the unselected state that https://github.com/material-components/material-components-ios/pull/1441 fixes.

DavidPhillipOster commented:
Interface Builder example added

ianegordon commented on components/Tabs/examples/TabBarViewControllerExample.m:
```diff
@@ -0,0 +1,79 @@
+/*
+ Copyright 2016-present the Material Components for iOS authors. All Rights Reserved.
```
nit: Confirm that this file is from 2016 and this is not a C&P error here and below.

DavidPhillipOster replied:
These files are from 2017. Thanks for the correction

ianegordon commented on components/Tabs/src/MDCTabBarViewController.m:
```diff
@@ -0,0 +1,226 @@
+
+- (void)setSelectedViewController:(nullable UIViewController *)selectedViewController {
+  if (_selectedViewController != selectedViewController) {
+#if DEBUG
```
Consider removing the #if DEBUG here.  NSAssert is disabled in RELEASE builds by default.

ianegordon commented on components/Tabs/src/MDCTabBarViewController.m:
```diff
@@ -0,0 +1,226 @@
+// Either this has just come into existence or its array of viewControllers has changed.
+// Update the TabBar from the array of viewControllers.
+- (void)updateTabBarItems {
+  NSMutableArray *items = [NSMutableArray array];
```
Add <UITabBarItem *> to array.

DavidPhillipOster replied:
done

ianegordon commented on components/Tabs/src/MDCTabBarViewController.m:
```diff
@@ -0,0 +1,226 @@
+  NSUInteger index = [tabBar.items indexOfObject:item];
+  if (index < _viewControllers.count) {
+   UIViewController *newSelected = _viewControllers[index];
+    if (newSelected != self.selectedViewController) {
```
indenting

DavidPhillipOster replied:
done

ianegordon reviewed (changes requested):
Examples demonstrated in person.

Commit:
Copyright date is 2017 for these TabBarViewController files. Fix white space issues.

Commit:
Add <UITabBarItem *> to local variable in updateTabBarItems

ianegordon commented:
Local scripts/test_all succeeded.

ianegordon commented:
Waiting on the +1 from Brian before Merge.

brianjmoore commented:
Please run clang-format against all the new files here:
```
scripts/install_contributor_tools
clang-format -style=file -i FILE
```

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,87 @@
+extern const CGFloat MDCTabBarViewControllerAnimationDuration;
+
+/**
+ The MDCTabBarViewController class manages a set of ViewControllers, showing their UITabBarItems
```
*view controllers

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ The MDCTabBarViewController class manages a set of ViewControllers, showing their UITabBarItems
+ as tappable items in a bottom MDCTabBar. When the user taps one of those items, the corresponding
+ ViewController appears.
```
*view controller

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,87 @@
+/** The tab bar controller's delegate. */
+@property(nonatomic, weak, nullable) id<MDCTabBarControllerDelegate> delegate;
+
+/** The array of view controllers managed by the tab bar controller.
```
nit here and below: first line of comment should be on its own line, like
```
/**
 The array of view controllers…
 */
```

DavidPhillipOster replied:
Done

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,87 @@
+
+ You can use this property to set colors but hide it, show it, and select
+ items in it using methods of this, MDCTabBarViewController, class.
+
```
remove extra newline?

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,87 @@
+ The tab bar which allows the user to switch between view controllers.
+
+ You can use this property to set colors but hide it, show it, and select
+ items in it using methods of this, MDCTabBarViewController, class.
```
omit "MDCTabBarViewController" here? I don't think it's needed

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.m:
```diff
@@ -0,0 +1,224 @@
+  }
+}
+
+- (void)setSelectedViewController:(nullable UIViewController *)selectedViewController {
```
Some brief comments breaking down the steps of this method might help readability a bit

DavidPhillipOster replied:
I don't understand what you are asking for here.

brianjmoore replied:
What I meant was that it takes a bit too much close reading to understand the high-level steps of this method: updating the property and showing the newly-visible view controller. (Similarly for the method above where it's not obvious that the first chunk of code is detaching removed child view controllers and the rest is updating the property). 

It's just about skimmability - Two brief comments separating out and denoting those two responsibilities would help future readers skim and understand the code more quickly.

DavidPhillipOster replied:
Done

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.m:
```diff
@@ -0,0 +1,224 @@
+    if (newSelected != self.selectedViewController) {
+      if ([_delegate respondsToSelector:@selector(tabBarController:shouldSelectViewController:)]
+          && ![_delegate tabBarController:self shouldSelectViewController:newSelected]) {
+        return;
```
Does this fulfill the contract that returning NO prevents the tab switch? I think the tab bar will still switch and leave things inconsistent

Possibly relevant - there is a tabBar:willSelectItem: that would be a good hook for the willSelectViewController: part of this API, but no shouldSelectItem.

DavidPhillipOster replied:
Modified  tabBar: to have shouldSelect, since it is a cover for UICollectionView's shouldSelect.

Commit:
Address Brian Moore's comments: ran clang-format -style=file on these files,
Modified tabBar delegate protocol: tabBar:willSelectItem: becomes tabBar:shouldSelectItem:
so that MDCTabBarViewController can have a 'shouldSelect' delegate.

Commit:
comment correction of ViewController to view controller.

DavidPhillipOster commented on components/Tabs/src/private/MDCItemBar.m:
```diff
@@ -1,5 +1,6 @@
 /*
```
I ran this file through clang-format, but my intent was that the only significant change was from willSelectItem to shouldSelectItem.

brianjmoore replied:
Can you double check your setup? It looks like the style from material-components-ios/.clang-format isn't being applied - the mangling below seems to be mostly due to wrapping to 80 columns

brianjmoore commented on components/Tabs/src/MDCTabBar.h:
```diff
@@ -141,9 +141,10 @@ IB_DESIGNABLE
+ If you don't implement all items changes are allowed.
  */
-- (void)tabBar:(nonnull MDCTabBar *)tabBar willSelectItem:(nonnull UITabBarItem *)item;
+- (BOOL)tabBar:(nonnull MDCTabBar *)tabBar shouldSelectItem:(nonnull UITabBarItem *)item;
```
Unfortunately this is an API change that would affect behavior so we have to follow the correct process to ensure we don't break clients.

What this means here:
* We have to add the new method rather than replacing the old one and implement a backward-compatible behavior so existing clients are unaffected.
* The new method needs an API review. I think we can do this inline here.
* The old method should be deprecated following the documented process. See https://github.com/material-components/material-components-ios/blob/develop/contributing/deprecation_policy.md

Commit:
Add new API to MDCItemBar in preparation for deprecating willSelect: from the delegate protocol, replacing it with shouldSelect:. Revert changes to comment lines that appear to be a mis-configuration at my end to 80 columns.

Commit:
Tested more thoroughly, verified that the class still works.

Commit:
Restore the delegate willSelect method that we intended in favor of shouldSelect:

brianjmoore commented on components/Tabs/src/private/MDCItemBar.h:
```diff
@@ -77,13 +77,21 @@
  selection changes.
  */
 @protocol MDCItemBarDelegate <NSObject>
+@optional
```
Since this is an internal, private interface let's be safe and keep everything required. You could actually delete willSelectItem from this protocol and just use shouldSelectItem to implement the MDCTabBar interface I think.

DavidPhillipOster replied:
Done

brianjmoore commented:
Left some inline comments above.

ajsecord reviewed (changes requested):
@ianegordon this seems to be missing the readme docs, the required files for the website generation, etc. Can we please land it into a feature branch until it's ready?

DavidPhillipOster commented on components/Tabs/src/private/MDCItemBar.h:
```diff
@@ -77,13 +77,21 @@
  selection changes.
  */
 @protocol MDCItemBarDelegate <NSObject>
+@optional
```
Done

Commit:
Simplify private/MDCItemBar's delegate protocol, write MDCTabBar public protocol in terms of it, as Brian asked.

Commit:
Comment fixes.

brianjmoore commented on components/Tabs/src/MDCTabBar.h:
```diff
@@ -141,7 +141,14 @@ IB_DESIGNABLE
+
+/**
+ Called before the selected item changes by user action. This method is not called for programmatic
+ changes to the tab bar's selected item.  TODO: Will be deprecated. Use tabBar:shouldSelectItem:.
```
Nit: "NOTE" instead of "TODO"

brianjmoore commented on components/Tabs/src/MDCTabBar.m:
```diff
@@ -224,11 +224,16 @@ - (void)itemBar:(MDCItemBar *)itemBar didSelectItem:(UITabBarItem *)item {
+    shouldSelect = [delegate tabBar:self shouldSelectItem:item];
   }
+  if (shouldSelect && [delegate respondsToSelector:@selector(tabBar:willSelectItem:)]) {
+      [delegate tabBar:self willSelectItem:item];
```
-2 spaces (clang-format should take care of this)

brianjmoore commented on components/Tabs/src/MDCTabBarViewController.h:
```diff
@@ -0,0 +1,89 @@
+
+ You can use this property to set colors but hide it, show it, and select
+ items in it using methods of this class.
+
```
remove extra newline?

brianjmoore commented on components/Tabs/src/private/MDCItemBar.m:
```diff
@@ -412,7 +426,8 @@ + (NSArray *)observableItemKeys {
     s_keys = @[
-      NSStringFromSelector(@selector(title)), NSStringFromSelector(@selector(image)),
+      NSStringFromSelector(@selector(title)),
+      NSStringFromSelector(@selector(image)),
```
Yeah, agreed that clang-format is wrong here. "//clang-format on/off" could help preserve the niceness or we could just accept the slight readability hit and welcome our new auto-formatting overlords.

DavidPhillipOster replied:
done

brianjmoore reviewed (approved):
Approved, pending a few inline comments and Adrian's request above. Thank you!

brianjmoore commented:
FYI: Bottom navigation support has landed, so you should be good to adopt it here. Should just need to implement -[UIBarPositioningDelegate positionForBar:] and return UIBarPositionBottom.

Commit:
Merge branch 'develop' of https://github.com/material-components/material-components-ios into MDCTabBarViewController

Commit:
Address Brian Moore's May 31 comments. Adjust color of TabBar in Interface builder example.

DavidPhillipOster commented:
• I believe I've addressed all of Brianjmoore's issues at this point. 
• Ajsecord asked for a readme file. I could use some help with that. 
• He also asked for 'the required files for the website generation' - I don't understand that at all - what specifically is missing?

ianegordon commented:
Note this PR got automagically close as part of the switching base branches.

It has landed in feature-tabcontroller.
